crates/core/src/metadata/mod.rs

```diff
@@ -96,6 +99,22 @@ impl Metadata {
         self.inner.nodes_config.load().version()
     }
 
+    /// Finds the effective leader node for a given partition. Use this method to discover the appropriate
+    /// worker node when routing requests to a partition.
+    pub fn get_partition_processor_node(&self, partition_id: PartitionId) -> Option<PlainNodeId> {
+        self.scheduling_plan_ref()
+            .get(&partition_id)
+            .and_then(|target_partition_state| target_partition_state.leader)
+    }
```

Comment:
Instead of adding this API to the `Metadata` type, can we separate it? That way, we don't require that the partition processor node information come from the metadata. The reason is that it's not fully clear that the `SchedulingPlan` will stay as it is right now and we might want to change the strategy how we obtain this information. One idea could be to introduce a `PartitionRoutingInformation` struct that exposes this API and is responsible for obtaining this information. Internally, it could fetch the information from the metadata store (e.g. periodically or whenever there is a signal to refresh the view).

Reply:
Makes sense, absolutely! I think a source of confusion for me was that the metadata manager / accessor API is called what it is, so I lumped it all together initially on the basis that the scheduling plan is saved in metadata store. Though that's somewhat different to other values we stash in there in that it's not the primary channel by which the plan is communicated to the worker nodes, it's mainly there for coordination among admin nodes as I understand it.

One could also make that case that we should have a single entry-point for runtime information access, be it for request routing, log access, or another purpose. It might make sense to have a single "runtime task context" that gives you access to these things are more (e.g. request identity in the future). Would be nice to avoid a proliferation of access mechanisms that essentially amount to the same thing - shared access to a snapshot of some locally-cached bit of config or state.

I won't make that case now though ;-) Happy to move this concern into a standalone `PartitionRoutingInformation` and take it from there.

Thanks again for the input!